File: Linked List/Stacks_Using_Linked_Lists/Stacks.py
class node:
    def __init__(self, value):
        self.data = value
        self.next = None

class Stacks:
    def __init__(self):
        self.TOS = None

    def Push(self, value):
        new_node = node(value)
        new_node.next = self.TOS
        self.TOS = new_node

    def Pop(self):
        if self.TOS == None:
            print("Stack is empty")
        else:
            self.TOS = self.TOS.next

    def Print_Stack(self):
        if self.TOS == None:
            print("Stack is empty")
        else:
            print("Top of stack is " + str(self.TOS.data))

            if self.TOS.next != None:
                print("Other elements are :", end = " ")
                current = self.TOS.next

                while current != None:
                    print(current.data, end = " ")
                    current = current.next

                print()

Stack = Stacks()

for i in range(0, 5):
    Stack.Push(i)

Stack.Print_Stack()

for i in range(0, 6):
    Stack.Pop()

Stack.Print_Stack()


''' Output

Top of stack is 4
Other elements are : 3 2 1 0
Stack is empty
Stack is empty

'''

File: Algo_Ds_Notes/Stacks_Using_Linked_Lists/Stacks.java
class Stacks
{
    node TOS;

    static class node
    {
        int data;
        node next;

        node(int value)
        {
            data = value;
        }
    }

    public void Push(int value)
    {
        node new_node = new node(value);
        new_node.next = TOS;
        TOS = new_node;
    }

    public void Pop()
    {
        if(TOS == null)
            System.out.println("Stack is empty");
        else
        {
            TOS = TOS.next;
        }
    }

    public void Print_Stack()
    {
        if(TOS == null)
            System.out.println("Stack is empty");
        else
        {
            System.out.println("Top of stack is " + TOS.data);

            if(TOS.next != null)
            {
                System.out.print("Other elements are : ");
                node current;
                current = TOS.next;

                while(current != null)
                {
                    System.out.print(current.data + " ");
                    current = current.next;
                }

                System.out.println();
            }
        }
    }

    public static void main(String[] args)
    {
        Stacks Stack = new Stacks();

        for(int i = 0; i < 5; i++)
            Stack.Push(i);

        Stack.Print_Stack();

        for(int i = 0; i < 6; i++)
            Stack.Pop();

        Stack.Print_Stack();
    }
}


/* Output

Top of stack is 4
Other Elements are : 3 2 1 0
Stack is empty
Stack is empty

*/
